### ISSUE TITLE
Subsonic api?

### ISSUE THREAD

**Issue by gerroon** [ampache/ampache#541]
Hi

The wiki says one can use subsonic apps but there is no info on how to do that. Can you please add a brief explanation on it?

thanks

**Comment by jeffkennedy5** [comment]
Same issue here. There doesn't seem to be a place to enter the Ampache url as a "web server URL" anywhere in the Subsonic GUI. If this URL should be in a configuration file somewhere instead, it would be nice to provide some kind of guidance on where that file can be located, or even a file name for the correct configuration file in SubSonic.

**Comment by Afterster** [comment]
First check that everything is working as describe in https://github.com/ampache/ampache/wiki/API#subsonic-api
Then you add your ampache url to the Server Address field in your Subsonic client, instead of "http://localhost:8080" you will put "http://localhost/ampache" for instance.

**Comment by jeffkennedy5** [comment]
I've gone through everything I can find on enabling the SubSonic API, but I still cannot get a successful session started on my local network.

I'm seeing the same issue as @gerroon described above: *"The file missing when I do "http://localhost/ampache/rest/ping.view" but I get the required message when I put instead "http://localhost/ampache/rest"*

When I go to URL *http://localhost/ampache/rest* I get the response of:

*"<subsonic-response version="1.10.1" status="failed">
    <error code="10" message="Required parameter is missing."/>
</subsonic-response>" *

But, when I go to URL *http://localhost/ampache/rest/ping.view," I get the following message:

*"Not Found

The requested URL /ampache/rest/ping.view was not found on this server."*

I'm sure that I am doing something wrong, but I can't figure out what it is. When I check the folder that should contain "ping.view" it is not in the folder. Complete contents of the Rest folder are as below:

.htaccess
.htaccess.dist
.htaccess.dist.bak
index.php

Now, I also have SubSonic installed on the same machine, and it is working fine on the default port of 4040. I was thinking that I might have a port issue, so, I've tried several different URLs in the SubSonic clients are:

http://localhost/ampache
http://localhost/ampache/rest
http://localhost/ampache/rest/
http://localhost:80/ampache
http://localhost:80/ampache/
http://localhost:80/ampache/rest
http://localhost:80/ampache/rest/

Any ideas on what might be wrong with my configuration?

**Comment by apastuszak** [comment]
Make sure mod_rewrite is enabled in your apache config.

**Comment by jeffkennedy5** [comment]
@apastuszak, Thanks. I can confirm that mod_rewrite is configured correctly on my Apache server, at least if the procedure at the following page is correct:
http://old.webasyst.net/support/help/general-modrewrite-test.html

I can also provide the contents of the .htaccess file located in the folder /www/ampache/rest/.htaccess:

*<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-s
    RewriteRule ^(.+)\.view$ /ampache/rest/index.php?action=$1 [PT,L,QSA]
    RewriteRule ^fake/(.+)$ /ampache/play/$1 [PT,L,QSA]
</IfModule>*

Any ideas on what I should look for next?

**Comment by jeffkennedy5** [comment]
oh... wait a minute.... I didn't have "/check" at the end of my URL to perform that test. So, maybe mod_rewrite isn't configured correctly after all, if it is supposed to still return the message *"mod_rewrite works!"*

The requested URL /ampache/mod_rewrite_test/check was not found on this server.

Do you know if that is the case?

**Comment by apastuszak** [comment]
Sorry.  Don't know enough about it to be able to answer your question.  It
works for me.

**Comment by jeffkennedy5** [comment]
@apastuszak, actually, you were right! I did some more digging and found the following link. I had to enable mod_rewrite with a command line of *a2enmod rewrite*

http://stackoverflow.com/questions/869092/how-to-enable-mod-rewrite-for-apache-2-2


After an Apache restart, this change is now at least giving me an error message in the client application of "The users permissions are not high enough." So, I'll need to track that down now, but I think I am getting closer.... will work with this some more to see if I can get final resolution. Thanks for identifying the fact that mod_rewrite wasn't working. You were exactly right about that.

**Comment by apastuszak** [comment]
I had the same exact issue when I set up my amapche server.  I think not
having mod_rewrite on is the #1 problem for the subsonic api not working.
Your current error may be related to not having api permissions enabled in
the ampache web console.

Go into the admin section of the ampache  console and add an ACL.

Set the ACL type to API/RPC.  Set the range from 0.0.0.0 to 255.255.255.255

Mine is set for "User: all" and Level All

**Comment by jeffkennedy5** [comment]
Yes! That did it... The existing ACL entries were not set for "Web Interface" and "Stream" only...I added an entry in Ampache, and it works in the iPhone AVSub app. Thanks for all of the help!

A few links that were also helpful to me in my particular situation, as I also found that .httaccess wasn't enabled, and I didn;t mention that earlier in this thread. 

http://stackoverflow.com/questions/9234289/verify-if-htaccess-file-is-running

**Comment by gerroon** [comment]
I still cant get this to work. I am using self signed certs and these certs are installed on my clients properly so that should not be an issue.

**Comment by Afterster** [comment]
If you get a 404 on ping.view then it's likely url rewriting is not working. Please double check mod_rewrite and .htaccess rewrite rules, the issue is coming from here for sure.

**Comment by gerroon** [comment]
Afterstep

Thanks for the reply, after messing around I found the issue. The .htaccess file in the rest folder must be copied to the root of Ampache folder and renamed to .htaccess. If it is at the root of the server then it has to be modified accordingly. I think that part should be made clear in the Wiki.

thanks

**Comment by Afterster** [comment]
The file under /rest/.htaccess shouldn't be copied on the root, otherwise it is a .htaccess / Apache misconfiguration. Copying to the root folder without adapting the file could result in some broken features.

**Comment by gerroon** [comment]
@Afterster 
keeping .htaccess under the rest folder does not work for me. I copied it to ampache root and it works. Not sure but the .htaccess file has /rest/... paths which makes me think that it is meant to be run from the root of the ampache folder

**Comment by Larx** [comment]
I ran into the same problem and was able to fix it according to this bug report. But that leaves one question for me: How is this .htaccess supposed to be created? 

When doing a regular install/update of Ampache according to the docs, the /rest-folder only contains .htaccess.dist, thus I had to manually create a .htaccess-file fitting for my installation path. There are also other folders containing only .htaccess.dist and not .htaccess (e.g. /bin), do I have to manually edit these files, too, in order not to break any functionality?

**Comment by gerroon** [comment]
@Larx 
how did you fix it?

**Comment by Larx** [comment]
I copies /rest/.htaccess.dist to /rest/.htaccess and modified the paths:

<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-s
    RewriteRule ^(.+)\.view$ /ampache/rest/index.php?ssaction=$1 [PT,L,QSA]
    RewriteRule ^fake/(.+)$ /ampache/play/$1 [PT,L,QSA]
</IfModule>

This enables Subsonic apps to connect to the server. However, something is still wrong, as I still cannot download or play the songs, but that might be some other configuration problem.

The incomplete .htaccess-files seem like a bug to me?!

**Comment by rZn** [comment]
Have you tried to disable stream_beautiful_url?

**Comment by Larx** [comment]
Yes. However, I currently cannot play any music at all with ampache, not only with the Subsonic API. There is something more wrong with my setup.

Concerning this bug: When  you read the docs, you nowhere find the hint that (at least) .htaccess.dist in the /rest folder has to be renamed and adapted to your needs. In contrary, following the current installation and upgrade notes, you definitely will not have the correct file in this place, as the distributed tarball does not contain a .htaccess-file.

I don't know if there are any other .htaccess.dist-files in the dir tree that have to be adapted, as this seems to be undocumented.

There seems to be at least a lack of documentation concerning installation and upgrade.

**Comment by Afterster** [comment]
Url rewriting stuff for streaming is under /play folder (/play/.htaccess in that case).
Normally you shouldn't have to do it manually as it is part of the installation process. This installation can fail to genereate the .htaccess files for your environment if write permission is missing. In that case it has to be done manually (populate the file lists at the end of the installation and download / copy them, or if you missed this step copy .htaccess.dist files to .htaccess and change local path if needed.
I will write something about that in the wiki Installation page.

**Comment by Larx** [comment]
A wiki entry would be really good. Even with complete permissions for the web server, I currently don't get any auto-generated .htaccess-files when updating according to the current instructions. Maybe only after a complete reinstall?

**Comment by Afterster** [comment]
During the update it should be done with update 360051. I don't know why here.
Quick Wiki entry about url rewriting: https://github.com/ampache/ampache/wiki/Installation#url-rewriting

**Comment by Larx** [comment]
Thanks, that helps.

**Comment by pravynandas** [comment]
Just in case if somebody still face this issue like me, added below note to Wiki:

* If not already done, you may need to copy /play/.htaccess.dist -> /play/.htaccess and /rest/.htaccess.dist -> /rest/.htaccess (and modify local paths as applicable)
